Create API for fence.register, and use it.

sofar opened the pull request:
This converts the call to minetest.register() for the default
fence node, so it can be called by other mods to quickly
setup other fences.

Since this creates an API, insert it into the game_api.txt.

The api looks like minetest.register(name, {def}), and has two
uncommon fields: "texture" and "material".

I'm not modifying the default fence recipe. Instead, this version
adds acacia sticks, pine sticks, junglewood sticks. This simplifies
the API and simplifies crafting, and is easier to remember, and
doesn't invent a new craft recipe that is hard to guess.

Lastly, adds pine, junglewood and acacia fences.

Textures for the new sticks. I verified all of minetest_game only
ever looks at group:stick for other material, so this prevents no
surprise to any player.

This is a from-scratch implementation, written by heart but inspired
by (#665 - Add many wooden fences).

The visual that fences connect strangely needs to be fixed in
content_mapblock.cpp, something that can easily be done.

sofar commented:
#665 - alternative implementation

Reason I'm posting this is not to dilute the conversation, but the better API implementation and removal of obscure recipes may make it more appealing.

paramat commented:
:+1: But i think for consistency the names should be:
default:stick_acacia_wood
default:stick_pine_wood
Since 'stick' should come first.

default:fence_acacia_wood
default:fence_pine_wood
Since we already have fence_wood and you are adding fence_junglewood.

sofar commented:
@paramat, agreed, default_stick*wood - is fine. Pushed an update implementing just that. Re-verified all the crafting recipes still work.

C1ffisme commented:
By the way, I would make the non apple wood sticks be ```not_in_creative_inventory```, because there is a very small chance that anyone will need to craft fences in creative mode, and it would be quite annoying to load an extra 3 items at startup (in creative mode).

sofar commented:
The one thing I don't like about not_in_creative_inventory, is that it breaks /give.

rubenwardy commented:
> The one thing I don't like about not_in_creative_inventory, is that it breaks /give.

Really? not_in_creative_inventory isn't mentioned in /give or /giveme: https://github.com/minetest/minetest/blob/master/builtin/game/chatcommands.lua#L518-L578

Do you mean that people with creative can't necessarily do /give on an item not in the inv?

sofar commented:
@rubenwardy nvm, I was incorrect. I had issues earlier testing some custom blocks and may have just mistyped.

PilzAdam commented:
:-1: a lot of different sticks just clutter the inventory and are generally annoying.

sofar commented:
On December 17, 2015 9:16:48 AM PST, PilzAdam <notifications@github.com> wrote:
>:-1: a lot of different sticks just clutter the inventory and are
>generally annoying.


I'll resubmit in a few days - I'm currently traveling a bit.

tenplus1 commented:
Agreed, ditch the different sticks and have it same as Ethereal fences with actual wood type on either side and normal sticks in the middle.

paramat commented:
> ditch the different sticks and have it same as Ethereal fences with actual wood type on either side and normal sticks in the middle.

I'm okay with that too, if other MTGame devs insist on it.

tenplus1 commented:
Returning six seems a little too much, how about 4 ?  We aready know that mc uses the new crafting recipe and gives only 3 which players are happy enough with.

kilbith commented:
:-1: Agree with @PilzAdam for the sticks.

sofar commented:
On December 18, 2015 3:15:04 AM GMT+01:00, Paramat <notifications@github.com> wrote:
>> ditch the different sticks and have it same as Ethereal fences with
>actual wood type on either side and normal sticks in the middle.
>
>I'm okay with that too, if other MTGame devs insist on it

Not a fan, at all

If you really want that, someone else should change the default fence recipe first. All fences should use the same recipe template, and ethereal breaks that pattern.

paramat commented:
> how about 4 ? We aready know that mc uses the new crafting recipe and gives only 3

Well never mind MC :} 6 from 18 sticks is consistent with our current recipe and is intuitive: 1 fence element being an upright plus 2 horizontals.

How about:
wood, wood, wood
wood, .......,  wood
The arch shape represents the form of the fence. I would prefer this.
5 x 4 = 20 = 18 + 2 sticks, return 18 / 3 = 6 fences.

paramat commented:
Or maybe:
wood, wood, wood
........, wood, .........
As this better represents the fence element with the upright central in a node.
4 x 4 = 16 sticks so return 15 / 3 = 5 fences.

C1ffisme commented:
Why not:
wood, wood, wood,
wood, wood, wood,

And change the trapdoor recipe to:

wood, stick, wood

Since trapdoors will most likely not be made for every wood type.

The trapdoor recipe also makes sense, because it has that cross made of sticks in the middle.

sofar commented:
@C1ffisme this patch has nothing to do with trapdoors, so let's not make it more complicated to get a change here by referring to a different block if it doesn't block a change here?

sofar commented:
@paramat I'm inclined to go with "wood-stick-wood" x2, since there are 2 cross-fence bars, and not 1. The alternative "T" shape craft seems a bit counter intuitive. I will have to modify the default fence recipe, which I had preferred to do in a separate commit.

sofar commented:
Old recipe was stickx6 -> 2 fences, which is 1.5 nodes -> 2 nodes, seems way too much to me to begin with.

I'd like to go with 4xwood+2x stick, which is 4.5 nodes -> 3 fences, a 1.5x reduction (0.66x)

sofar commented:
- Removed sticks
- changed recipe to "wood, stick, wood" x2
- returns 3 fences
- change "wood" to "acacia wood" to craft acacia fence

I believe this addresses all comments listed above. Can I have some "thumbs up" now? :^)

paramat commented:
> Old recipe was stickx6 -> 2 fences, which is 1.5 nodes -> 2 nodes, seems way too much to me to begin with.

Makes sense.
I prefer 4.5 nodes -> 4 fences.

sofar commented:
- recipe changed to output 4 fences instead of 3.

Since @tenplus1 and @paramat requested this, and I'm not too attached to the output amount, I was happy to change it.

Any more nitpicks ? ;^)

paramat commented:
:+1:

C1ffisme commented:
> recipe changed to output 4 fences instead of 3.

Thank goodness, fences are expensive. :) 

:+1:

Fixer-007 commented:
+1

HybridDog commented on mods/default/functions.lua:
```diff
@@ -210,6 +210,37 @@ end
+		output = name .. " 4",
+		recipe = {
+			{ def.material, 'group:stick', def.material },
+			{ def.material, 'group:stick', def.material },
```
it's the recipe of the other game
http://minecraft-de.gamepedia.com/Zaun#Herstellung

sofar replied:
read the discussion above, this was changed on purpose.

HybridDog replied:
>read the discussion above […]

sorry

sofar replied:
np!

sofar commented:
It would help if some folks that have expressed concerns before could re-review.

@pilzadam @kilbith @tenplus1 @rubenwardy

PilzAdam commented:
What happens when I have 2 wood from one type, and two from another? Can I create a fence out of it?

sofar commented:
@pilzadam, no, as that would require many recipes to implement, unless you make a kludge recipe where only one of the wood blocks is not group:wood in a specific spot, but that sounds very confusing.

I'm aware that this may be inconvenient, but wood isn't exactly extremely hard to come by.

PilzAdam commented:
@sofar I don't like that it behaves like this. Could it be possible to add a "fallback" recipe with only group:wood that creates the normal fence?

sofar commented:
@pilzadam I'll investigate. I suspect that we can't add a fallback recipe with entirely "group:wood" as I think it's undefined whether you'll be able to make acacia fences if the core sorts the recipes wrongly, but maybe we can find a creative solution somewhere.

paramat commented:
> Could it be possible to add a "fallback" recipe with only group:wood that creates the normal fence?

That's messy and unnecessary, only 2 trunk nodes are needed to complete the recipe so there's no need to be able to mix woods. Getting applewood fence from junglewood / acacia / pine will seem wrong.

sofar commented:
@paramat the current recipe is: "wood, stick, wood" x2. I think you're confused with the MC "gate" recipe.

However, I agree with you that any fallback will be confusing in itself and will likely have a negative impact as well as a positive one. (player: "why am I getting an apple fence, I clearly have 4 pine wood,... oh wait I have 3 pine wood and one apple wood" - type situations).

sofar commented:
Adding aspen wood fences into the mix, as Aspen are now merged.

sofar commented:
Adding the following fallback recipe
```
group:wood stick group:wood
group:wood stick group:wood
```
would make all the other fences uncraftable, as every fence crafted will be a ```default:fence_wood```

sofar commented:
I'm proposing no more changes, and no inclusion of fallback recipes, as is the last version of this patch.

HybridDog commented:
>Adding the following fallback recipe

>group:wood stick group:wood
group:wood stick group:wood
>
would make all the other fences uncraftable, as every fence crafted will be a default:fence_wood

The crafting code at builtin can be edited that craft recipes without groups are preferred (tested first and aborted), can't it?

sofar commented:
@HybridDog I think that's not done in lua, but in C++ instead.

I'm open to the idea, but I think it's separate from this patch - and should not preclude it.

paramat commented:
:+1:

Fixer-007 commented:
+1

0-afflatus commented:
I'm catching up after being offline for a month or so.
The code looks good. I'm wary of the obfuscation potential of APIs (c.f. buckets), but I think you have avoided that. Thank you for not proliferating sticks, I'll have a go at breaking it.

ShadowNinja commented on mods/default/functions.lua:
```diff
@@ -210,6 +210,37 @@ end
+		},
+		groups = def.groups,
+		sounds = def.sounds,
+	})
```
Instead of just copying over certain spcific fields, how about just adding the needed fields to the definition?  This way you'll be able to do things like add glowglass fences.

sofar replied:
Thanks, this was bothering me as well and I should have adjusted it earlier. I've made almost everything entirely optional, adjusted docs and at this point one can pass through any nodedef property.

ShadowNinja commented on mods/default/nodes.lua:
```diff
@@ -151,6 +151,10 @@ default:sign_wall
 
 default:fence_wood
+default:fence_acacia_wood
+default:fence_junglewood
```
`jungle_wood`?

paramat replied:
For backwards compatibility and simplicity we decided to keep the name unchanged, pine_wood and acacia_wood have the new 2-word form because they are new.

ShadowNinja replied:
This is new though, no B/C required.

sofar replied:
Right, but slabs and stairs are also ```junglewood``` and not ```jungle_wood```. Hence Introducing one node that is off seemed confusing.

paramat replied:
Yes junglewood is better for name consistency.

sofar commented:
Added the ability to make pass virtually any nodedef property through so one can make full use of the nodedef api, including event handlers and such, as noted by @shadowninja.

I also added an explicit  ```groups.fence =1```

ShadowNinja commented on mods/default/functions.lua:
```diff
@@ -210,6 +210,61 @@ end
+	if not def.groups then
+		def.groups = {}
+	end
+	if def.groups then
```
This is necessarily true, because of the previous block (unless you're doing something screwy like overriding the `def` metatable's `__newindex` of course, which we don't support).

ShadowNinja commented on mods/default/functions.lua:
```diff
@@ -210,6 +210,61 @@ end
+			type = "fixed",
+			fixed = {-1/7, -1/2, -1/7, 1/7, 1/2, 1/7},
+		}
+	end
```
Here's a neater way to do this:
```
local default_fields = {
	paramtype = "light",
	<...>
}
for k, v in pairs(default_fields) do
	if not def[k] then
		def[k] = v
	end
end

ShadowNinja commented on mods/default/functions.lua:
```diff
@@ -210,6 +210,61 @@ end
+	local fence_texture = "default_fence_overlay.png^" .. def.texture ..
+			"^default_fence_overlay.png^[makealpha:255,126,126"
+	def.drawtype = "fencelike"
+	-- allow almost everything to be overridden
```
Comments should follow English capitalization rules -- which means starting with a capital letter.

ShadowNinja commented:
Few other minor issues, otherwise :+1:.

sofar commented:
Updated with all of @ShadowNinja 's comments, thanks.

paramat commented:
Nice, will merge soon.

paramat commented:
2f39cad09b763fc62dce3bf1dce2867ecc864955

HybridDog commented on mods/default/functions.lua:
```diff
@@ -210,6 +210,51 @@ end
+		groups = {},
+	}
+	for k, v in pairs(default_fields) do
+		if not def[k] then
```
def[k] == nil
